### ISSUE TITLE
Support for pet water dispenser mmgg.pet_waterer (Xiaowan XWWF01MG)

### ISSUE THREAD

**Issue by gudvinr** [rytilahti/python-miio#897]
**Device information:**

  - Name(s) of the device: Xiaowan Smart Pet Water Dispenser
  - Link: [AliExpress](https://www.aliexpress.com/i/4001143786885.html)
  - Model: `mmgg.pet_waterer.s1` and `mmgg.pet_waterer.s4`
  - Hardware version: `esp8266` (s1)
  - Firmware version: `2.0.4_0006`

**Additional context**
s1 and s4 version almost identical according to miot spec except s4 has one additional property. s4 probably is an EU version of this water dispenser and s1 is Chinese one.

miot-spec:
* [mmgg-s1.json](https://github.com/rytilahti/python-miio/files/5748859/urn.miot-spec-v2.device.pet-drinking-fountain.0000A067.mmgg-s1.1.json.txt)
* [mmgg-s4.json](https://github.com/rytilahti/python-miio/files/5748858/urn.miot-spec-v2.device.pet-drinking-fountain.0000A067.mmgg-s4.1.json.txt)

**Comment by HydrelioxGitHub** [comment]
I'm interested as I acquired this device a couple of months ago. If I could help in any way ..

**Comment by rytilahti** [comment]
If you want to contribute, please take a look at other miot platforms (and `devtools/` directory for miottemplate to make spec files easier to comprehend).

**Comment by HydrelioxGitHub** [comment]
I followed the guide provided by @gudvinr and I have been able to send commands to the device through command line.

I have take a quick look at other devices. I think I could do something but I'm not very skilled so it will take some time. Maybe waterpurifier.py is the most useful example I will take for example (work in a close way as water dispenser, simple stucture).

**Comment by HydrelioxGitHub** [comment]
I do not get where python-miio do the match between property name written in the code and siid, piid and so on ...

**Comment by HydrelioxGitHub** [comment]
I saw by reading some comment that @gudvinr is much more skilled than me. Is he still working on this issue ?

**Comment by rytilahti** [comment]
@HydrelioxGitHub take a look at https://github.com/rytilahti/python-miio/blob/master/miio/heater_miot.py (or any other `_miot.py` file) for some ideas, I think the heater one is quite straightforward here so that's why I'm linking it.

The `_MAPPING` in the beginning links the name of the variable (that's passed as `did` for later matching) to the piid and siid (take a look at `__init__` which passed the mapping to the MiotDevice constructor so that methods like `set_property` work properly, and `status()` which uses the mapping (via `get_properties_for_mapping`) for the status request).

**Comment by gudvinr** [comment]
@rytilahti this device need `actions` support for filter reset though

**Comment by rytilahti** [comment]
@gudvinr ah, I see. I think support for actions should be added to #905, I'll check it out when I find some free time to do that.

**Comment by rezmus** [comment]
for some reason some actions can't be called via miio udp 54321, only via /miotspec/action cloud api.

**Comment by rytilahti** [comment]
@rezmus do you have examples of such behavior? What happens if you try to call them, could the problem be just missing input parameters?

**Comment by rezmus** [comment]
for example viomi se robot (viomi.vacuum.v19), clean room id "10"

{"method":"action","params":{"did":"X","siid":4,"aiid":13,"in":[0,1,"10"]},"id":123}

if you send this to robot via miio resonse is code 0, but robot only makes beep and refuse to clean. if you send such command via cloud api /miotspec/action works just fine. i also remember it refuse to download sound package via miio which was similar action with "in" array contains voice pack params (url, md5, id).

however actions with empty "in" are executed fine via local miio for example start full cleanup

{"method":"action","params":{"did":"X","siid":2,"aiid":7,"in":[]},"id":123}

**Comment by al-one** [comment]
https://github.com/al-one/hass-xiaomi-miot/commit/10e849490d8e1abe15c913449b42840684fd4ca2

**Comment by HydrelioxGitHub** [comment]
Hello @al-one , What feature can we get with your integration ? I suppose switch on and off ? Sensor on properties ? I will try this as soon as I can.

**Comment by al-one** [comment]
> Hello @al-one , What feature can we get with your integration ? I suppose switch on and off ? Sensor on properties ? I will try this as soon as I can.

You will get:
- **switch** for power
- **fan** for mode
- **sensor** for filter

**Comment by HydrelioxGitHub** [comment]
> > Hello @al-one , What feature can we get with your integration ? I suppose switch on and off ? Sensor on properties ? I will try this as soon as I can.
> 
> You will get:
> - **switch** for power
> - **fan** for mode
> - **sensor** for filter

Thanks for your time.

**Comment by HydrelioxGitHub** [comment]
I added my water dispenser token and ip to Xiaomi Miot Auto but I only have one sensor for the filter time before to change, no switch, no fan ...
But this is not really related to this issue, maybe I should open another one on your repository .

**Comment by al-one** [comment]
Download last commit of master.

**Comment by HydrelioxGitHub** [comment]
> Download last commit of [master](https://github.com/al-one/hass-xiaomi-miot).

I'm using HACS and I don't want to mess with something so I will wait for the next release.

**Comment by al-one** [comment]
> I'm using HACS and I don't want to mess with something so I will wait for the next release.

Support pet-drinking-fountain after [v0.2.6](https://github.com/al-one/hass-xiaomi-miot/releases/latest).

**Comment by roberdlv** [comment]
> > I'm using HACS and I don't want to mess with something so I will wait for the next release.
> 
> Support pet-drinking-fountain after [v0.2.6](https://github.com/al-one/hass-xiaomi-miot/releases/latest).

I am using your integration for Pet water xiaomi, I get switch, fan, sensor and light, I added each one in Entities in my yaml, but the functions are not correct, am I doing something wrong? light does nothing, sensor does mark, but switch and fan only turn it on or off, it does not change on the smart and normal modes. Thank you!

**Comment by al-one** [comment]
> I am using your integration for Pet water xiaomi, I get switch, fan, sensor and light, I added each one in Entities in my yaml, but the functions are not correct, am I doing something wrong? light does nothing, sensor does mark, but switch and fan only turn it on or off, it does not change on the smart and normal modes. Thank you!

Integrate by yaml is not recommended. If you have to use yaml, just configure switch:
```yaml
switch:
  - platform: xiaomi_miot
    host: 192.168.lan.ip
    token: xxxxxx
    model: mmgg.pet_waterer.s1
```

**Comment by roberdlv** [comment]
> > I am using your integration for Pet water xiaomi, I get switch, fan, sensor and light, I added each one in Entities in my yaml, but the functions are not correct, am I doing something wrong? light does nothing, sensor does mark, but switch and fan only turn it on or off, it does not change on the smart and normal modes. Thank you!
> 
> Integrate by yaml is not recommended. If you have to use yaml, just configure switch:
> 
> ```yaml
> switch:
>   - platform: xiaomi_miot
>     host: 192.168.lan.ip
>     token: xxxxxx
>     model: mmgg.pet_waterer.s1
> ```

So how could I add it to my HA interface? I already see the entities of light, switch, fan and sensor.

**Comment by roberdlv** [comment]
![Screenshot_2](https://user-images.githubusercontent.com/56409767/113861354-6a149b00-97a7-11eb-9393-096d34f0c8c4.jpg)

**Comment by HydrelioxGitHub** [comment]
> > > I am using your integration for Pet water xiaomi, I get switch, fan, sensor and light, I added each one in Entities in my yaml, but the functions are not correct, am I doing something wrong? light does nothing, sensor does mark, but switch and fan only turn it on or off, it does not change on the smart and normal modes. Thank you!
> > 
> > Integrate by yaml is not recommended. If you have to use yaml, just configure switch:
> > 
> > ```yaml
> > switch:
> >   - platform: xiaomi_miot
> >     host: 192.168.lan.ip
> >     token: xxxxxx
> >     model: mmgg.pet_waterer.s1
> > ```
> 
> So how could I add it to my HA interface? I already see the entities of light, switch, fan and sensor.

You should go in Configuration->Integration then use the Add Integration in the bottom right corner. Then search for "Xiaomi Miot Auto" and click on it and just fill the form.

**Comment by roberdlv** [comment]
> > > > Estoy usando su integración para Pet water xiaomi, obtengo interruptor, ventilador, sensor y luz, agregué cada uno en Entidades en mi yaml, pero las funciones no son correctas, ¿estoy haciendo algo mal? la luz no hace nada, el sensor marca, pero el interruptor y el ventilador solo lo encienden o apagan, no cambia en los modos inteligente y normal. ¡Gracias!
> > > 
> > > 
> > > No se recomienda la integración de yaml. Si tiene que usar yaml, simplemente configure el interruptor:
> > > ```yaml
> > > cambiar :
> > >   - plataforma : xiaomi_miot 
> > >     host : 192.168.lan.ip 
> > >     token : xxxxxx 
> > >     modelo : mmgg.pet_waterer.s1
> > > ```
> > 
> > 
> > Entonces, ¿cómo podría agregarlo a mi interfaz HA? Ya veo las entidades de luz, interruptor, ventilador y sensor.
> 
> Debe ir a Configuración-> Integración y luego usar Agregar integración en la esquina inferior derecha. Luego busque "Xiaomi Miot Auto" y haga clic en él y simplemente complete el formulario.

Sí, exactamente es eso lo que hice, pero para añadirlo en la interface de HA tuve que añadir como Entidades una por una, entonces que fan / switch y luz no tienen funcion correcta, switch y fan solo encieden y apagan pero no eligen el modo normal o inteligente, luz no hace nada, sensor si marca correctamente los días restantes.

![Screenshot_3](https://user-images.githubusercontent.com/56409767/113862094-443bc600-97a8-11eb-8401-03345f79480c.jpg)

**Comment by gudvinr** [comment]
@al-one @roberdlv I suggest you to file an issue in hass-xiaomi-miot repo where it belongs. This issue was filled to support device in python-miio and it is kinda annoying to receive notifications related to bugs in another software.

@rytilahti can you mark unrelated messages as off topic please?

**Comment by roberdlv** [comment]
> @al-one @roberdlv I suggest you to file an issue in hass-xiaomi-miot repo where it belongs. This issue was filled to support device in python-miio and it is kinda annoying to receive notifications related to bugs in another software.
> 
> @rytilahti can you mark unrelated messages as off topic please?

Ok Manolo, so it will be!

**Comment by ofen** [comment]
https://github.com/rytilahti/python-miio/pull/1174

**Comment by OXOTH1K** [comment]
Guys, can someone dump stock firmware from water dispenser? My FW was broken( A little guide can be found [here](https://github.com/mmakaay/esphome-xiaomi_bslamp2/blob/dev/doc/flashing.md#make-a-backup-of-the-current-firmware).

**Comment by gudvinr** [comment]
> Guys, can someone dump stock firmware from water dispenser? My FW was broken( A little guide can be found [here](https://github.com/mmakaay/esphome-xiaomi_bslamp2/blob/dev/doc/flashing.md#make-a-backup-of-the-current-firmware).

**Caution**: You will find the WLAN SSID and Password of the last used WiFi network in this file. Therefore, keep this backup in a safe place.

That doesn't look quite safe.

**Comment by OXOTH1K** [comment]
> That doesn't look quite safe.
You can reset device to default settings. It will erase config for WiFI.

**Comment by mSapik** [comment]
> > Hello @al-one , What feature can we get with your integration ? I suppose switch on and off ? Sensor on properties ? I will try this as soon as I can.
> 
> You will get:
> 
> * **switch** for power
> * **fan** for mode
> * **sensor** for filter

Hello, can you add a support for a 'no-water-flag' mmgg-s1 model? Like a binary sensor.

>   {
>       "iid": 7,
>       "type": "urn:mmgg-spec:service:no-water-flag:00007803:mmgg-s1:1",
>       "description": "",
>       "properties": [
>         {
>           "iid": 1,
>           "type": "urn:mmgg-spec:property:no-water-flag:00000001:mmgg-s1:1",
>           "description": "0：缺水 1：有水",
>           "format": "bool",
>           "access": [
>             "read",
>             "notify"
>           ],
>           "unit": "none"
>         },
>         {
>           "iid": 2,
>           "type": "urn:mmgg-spec:property:no-water-time:00000002:mmgg-s1:1",
>           "description": "",
>           "format": "uint16",
>           "access": [
>             "read",
>             "notify"
>           ],
>           "unit": "minutes",
>           "value-range": [
>             0,
>             10080,
>             1
>           ]
>         }

**Comment by gudvinr** [comment]
@rytilahti I'll close this issue since support is merged in #1174 and already in master since 0.5.9

**Comment by D18x** [comment]
I've bought anXiaomi Smart Pet Fountain 2, but i can't add it to the Xiaomi App, unless i set the application region to China MainLand.
Where can i get the EU firmware version?
